File: library/src/main/java/zjm/cst/dhu/library/utils/CacheInterceptor.java
package zjm.cst.dhu.library.utils;

import java.io.IOException;

import okhttp3.CacheControl;
import okhttp3.Interceptor;
import okhttp3.Request;
import okhttp3.Response;
import zjm.cst.dhu.library.MyApplication;

/**
 * Created by zjm on 2017/5/10.
 */

public class CacheInterceptor implements Interceptor {
    @Override
    public Response intercept(Chain chain) throws IOException {
        Request request = chain.request();
        if (!NetworkUtil.isNetworkConnected(MyApplication.getContext())) {
            request = request.newBuilder()
                    .cacheControl(CacheControl.FORCE_CACHE)
                    .build();
        }
        Response response = chain.proceed(request);
        if (NetworkUtil.isNetworkConnected(MyApplication.getContext())) {// 有网络时 设置缓存超时时间0个小时
            int maxAge = 0 * 60;
            response.newBuilder()
                    .header("Cache-Control", "public, max-age=" + maxAge)
                    .removeHeader("Pragma") // 清除头信息，因为服务器如果不支持，会返回一些干扰信息，不清除下面无法生效
                    .build();
        } else { // 无网络时，设置超时为2天
            int maxStale = 60 * 60 * 24 * 2;
            response.newBuilder()
                    .header("Cache-Control", "public, only-if-cached, max-stale=" + maxStale)
                    .removeHeader("Pragma")
                    .build();
        }
        return response;
    }
}

File: library/src/main/kotlin/zjm/cst/dhu/library/utils/network/CacheInterceptor.kt
package zjm.cst.dhu.library.utils.network

import zjm.cst.dhu.library.Constant
import zjm.cst.dhu.library.MyApplication
import zjm.cst.dhu.library.utils.network.NetworkUtil
import okhttp3.CacheControl
import okhttp3.Interceptor
import okhttp3.Response
import java.io.IOException

/**
 * Created by zjm on 2017/5/10.
 */

class CacheInterceptor
    : Interceptor {

    @Throws(IOException::class)
    override fun intercept(chain: Interceptor.Chain):
            Response {
        var request = chain.request()
        if (!NetworkUtil.isNetworkConnected(MyApplication.context)) {
            request = request.newBuilder()
                    .cacheControl(CacheControl.FORCE_CACHE)
                    .build()
        }
        val response = chain.proceed(request)
        if (NetworkUtil.isNetworkConnected(MyApplication.context)) {// 有网络时 设置缓存超时时间0个小时
            val maxAge = Constant.CACHE_NETWORK
            response.newBuilder()
                    .header("Cache-Control", "public, max-age=" + maxAge)
                    .removeHeader("Pragma") // 清除头信息，因为服务器如果不支持，会返回一些干扰信息，不清除下面无法生效
                    .build()
        } else { // 无网络时
            val maxStale = Constant.CACHE_NONETWORK
            response.newBuilder()
                    .header("Cache-Control", "public, only-if-cached, max-stale=" + maxStale)
                    .removeHeader("Pragma")
                    .build()
        }
        return response
    }
}
